File: apps/api/src/models/questionnaire.ts
import {prisma} from './client';

export type QuestionType = 'single' | 'multi';

export type QuestionnaireOption = {
  id: string;
  label: string;
  score?: number;
};

export type QuestionnaireStep = {
  id: string;
  type: QuestionType;
  eyebrow: string;
  question: string;
  hint: string;
  minSelections?: number;
  maxSelections?: number;
  options: QuestionnaireOption[];
};

export type Questionnaire = {
  id: string;
  version: number;
  title: string;
  description: string;
  steps: QuestionnaireStep[];
};

type QuestionnaireRecord = {
  id: string;
  version: number;
  title: string;
  description: string;
  steps: {
    stepId: string;
    type: string;
    eyebrow: string;
    question: string;
    hint: string;
    minSelections: number | null;
    maxSelections: number | null;
    options: {
      optionId: string;
      label: string;
      score: number | null;
    }[];
  }[];
};

function fromRecord(record: QuestionnaireRecord): Questionnaire {
  return {
    id: record.id,
    version: record.version,
    title: record.title,
    description: record.description,
    steps: record.steps.map((step) => ({
      id: step.stepId,
      type: step.type as QuestionType,
      eyebrow: step.eyebrow,
      question: step.question,
      hint: step.hint,
      ...(step.minSelections === null ? {} : {minSelections: step.minSelections}),
      ...(step.maxSelections === null ? {} : {maxSelections: step.maxSelections}),
      options: step.options.map((option) => ({
        id: option.optionId,
        label: option.label,
        ...(option.score === null ? {} : {score: option.score}),
      })),
    })),
  };
}

export const QuestionnaireModel = {
  async findActive(): Promise<Questionnaire | null> {
    const questionnaire = await prisma.questionnaire.findFirst({
      where: {isActive: true},
      orderBy: {updatedAt: 'desc'},
      include: {
        steps: {
          orderBy: {position: 'asc'},
          include: {
            options: {
              orderBy: {position: 'asc'},
            },
          },
        },
      },
    });

    return questionnaire ? fromRecord(questionnaire) : null;
  },

  async seed(questionnaire: Questionnaire) {
    await prisma.$transaction(async (tx) => {
      await tx.questionnaire.upsert({
        where: {id: questionnaire.id},
        create: {
          id: questionnaire.id,
          version: questionnaire.version,
          slug: questionnaire.id,
          title: questionnaire.title,
          description: questionnaire.description,
          isActive: true,
        },
        update: {
          version: questionnaire.version,
          slug: questionnaire.id,
          title: questionnaire.title,
          description: questionnaire.description,
          isActive: true,
        },
      });

      await tx.questionnaireStep.deleteMany({
        where: {questionnaireId: questionnaire.id},
      });

      for (const [stepIndex, step] of questionnaire.steps.entries()) {
        await tx.questionnaireStep.create({
          data: {
            questionnaireId: questionnaire.id,
            stepId: step.id,
            position: stepIndex,
            type: step.type,
            eyebrow: step.eyebrow,
            question: step.question,
            hint: step.hint,
            minSelections: step.minSelections,
            maxSelections: step.maxSelections,
            options: {
              create: step.options.map((option, optionIndex) => ({
                optionId: option.id,
                position: optionIndex,
                label: option.label,
                score: option.score,
              })),
            },
          },
        });
      }
    });
  },
};

File: apps/api/src/models/questionnaire.py
from datetime import UTC, datetime

from sqlalchemy import UniqueConstraint
from sqlmodel import Field, Relationship, SQLModel


class QuestionRecord(SQLModel, table=True):
    __tablename__ = "questions"
    __table_args__ = (UniqueConstraint("position", name="uq_questions_position"),)

    id: str = Field(primary_key=True)
    is_active: bool = Field(default=True, index=True)
    position: int
    weight: float = Field(default=1)
    type: str
    eyebrow: str
    question: str
    hint: str
    min_selections: int | None = None
    max_selections: int | None = None
    options: list["QuestionOptionRecord"] = Relationship(
        back_populates="question",
        sa_relationship_kwargs={"order_by": "QuestionOptionRecord.position"},
    )
    created_at: datetime = Field(default_factory=lambda: datetime.now(UTC))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(UTC))


class QuestionOptionRecord(SQLModel, table=True):
    __tablename__ = "question_options"
    __table_args__ = (
        UniqueConstraint(
            "question_id",
            "option_id",
            name="uq_question_options_question_option",
        ),
        UniqueConstraint(
            "question_id",
            "position",
            name="uq_question_options_question_position",
        ),
    )

    id: int | None = Field(default=None, primary_key=True)
    question_id: str = Field(foreign_key="questions.id", ondelete="CASCADE", index=True)
    option_id: str = Field(index=True)
    position: int
    label: str
    score: int | None = None
    question: QuestionRecord | None = Relationship(back_populates="options")
    created_at: datetime = Field(default_factory=lambda: datetime.now(UTC))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(UTC))
